third_party/nvfuser/csrc/python_frontend/python_bindings.cpp

```diff
@@ -96,20 +96,17 @@ void initNvFuserPythonBindings(PyObject* module) {
           py::arg("max_length") = int(1024))
       .def_readwrite("ops", &nvfuser::FusionDefinition::ops)
       .def(
-          "__enter__",
+          "_setup_definition",
```

Comment:
do we need the underscore to differentiate the pybind entry from the cpp method?

Reply:
I understand that we are trying to "hide" pybind entries. Since we are already exposing them via `nvfuser._C.xxx`, this feels like a little redundant.

Reply:
No, the point of the underscore in the method names is that python treats leading underscore names specially as hidden methods that are meant to be internal kind of like private methods in C++ without the protections.  So, given that `FusionDefinition` will inherit from `_FusionDefinition`, an underscore method like `_setup_definition` is still callable but it isn't intended to be user facing and part of the API for general use.  If you do `help` on the class you should not see the method.

Reply:
By adding the leading underscore to the python API names, I am trying to indicate the methods are intended for internal use only.

Reply:
Ha got'ya. I forgot that `FusionDefinition` inherits from `_FusionDefinition`. I'm still treating that as a wrapper.

A side note, why do we need to inherit from from `_FusionDefinition`, but not a wrapper instead? I think the intention is to separate what we expose in cpp/pybind and have a more organized and user friendly python API. So separating what python `FusionDefinition` from the cpp object seems like a good idea?

Granted, that initially we'll have some boring boilerplate wrapper method that directs to the cpp methods... We can also just do that in later PR, if it makes sense.

Reply:
In case I'm not being clear. ^^ this is just my opinion. we don't necessarily have to go that way with separating python / pybind APIs.

Reply:
For the definition side, the C++ `FusionDefinition` is not just a wrapper around the Fusion IR Arith functions.  Prior to adding `FusionCache`, it was more or less a wrapper as the `Operators` directly called arith functions in the Fusion IR.  Now, the `_FusionDefinition` is holding the Record representation of the definition.  Prior to the `_FusionDefinition` holding state, I think you are right that it probably made more sense to just wrap arith functions.  The limitation there was we didn't have a good way to store python files under our TorchScript home, so it was easier to just manage all of the python API through the pybindings.

So, the reason for using a Python `FusionDefinition` and not just the `_FusionDefinition` is that it code introspection is easier in Python when dealing with parent and child classes specifically in managing the definition as a function instead of just as a context manager.  The other reason is that the error messages are less cryptic for the context manager.  When the user provided garbage to the context manager with the C++ pybindings, it was getting flagged in the `__exit__()` method instead of as a more natural python error like an unknown name.

Reply:
I will note, I didn't use code introspection, yet.  However, I intend to query whether subclass methods are implemented to conditionalize execution paths under the python `execute()` method.

Reply:
sorry my message come out wrong. I wasn't questioning the cpp side APIs at all. And I'm all in for having a Python `FusionDefinition`.

The comment I have is on the decision what we have in `third_party/nvfuser/python/__init__.py`

`class FusionDefinition(_C._FusionDefinition):`

Do we want it to be an inheritance here, instead of a wrapper? I agree it's easier having it that way instead of plumbing all the pybind APIs. But I thought that's opposite to our intention here, which is to separate python and cpp APIs.

Reply:
I think I am having a failure of imagination.  What would a wrapper look like instead of inheriting `_C._FusionDefinition`?

The reason I am using inheritance is that I don't want to redefine all class methods in the `python` class.